File: WhatsappDecrypt.py
import sys
import os
import time
import binascii
from jeb.api import IScript
from jeb.api import EngineOption
from jeb.api.ui import View
from jeb.api.dex import Dex
from jeb.api.ast import Class, Field, Method, Call, Constant, StaticField, NewArray

class WhatsappDecrypt(IScript):
  
  def decode_string(self,encoded_str, key):
   decoded_str = ''
   for i in range(len(encoded_str)):
    decoded_str += chr(ord(encoded_str[i]) ^ key[i % 5])
   return decoded_str

  def run(self, jeb):
    self.keys = [0x29,0x49,0x70,0x1C,0x71] #Change keys here (see switch statements)
    self.jeb = jeb
    self.dex = self.jeb.getDex()
    if not self.dex:
      print 'Error! Please provide an input file.'
      sef.jeb.exit()

    v = self.jeb.getUI().getView(View.Type.JAVA)
    if not v:
      print 'Switch to the Java view, position the caret somewhere inside the method to be decompiled'
      return
    self.msig=v.getCodePosition().getSignature()
    print(self.msig)
	
    md = self.dex.getMethodData(self.msig)
    if not md:
     print ("Failed get Method Data")
     return
    
    code = md.getCodeItem()
    if not code:
     return
    
    stringlist = {}
    for insn in code.getInstructions():
     if not insn.getMnemonic() in ('const-string', 'const-string/jumbo'):
      continue
     stringindex = insn.getParameters()[1].getValue()
     s = self.dex.getString(stringindex)
     s2 = self.decode_string(s,self.keys)
     if (stringindex not in stringlist):
      stringlist[stringindex]=s2
    for m in stringlist:
     print (hex(m)+": "+stringlist[m])
     self.dex.setString(m,stringlist[m])
    print("Done")
    return

File: WhatsAppDecrypt.java
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import jeb.api.IScript;
import jeb.api.JebInstance;
import jeb.api.dex.Dex;
import jeb.api.dex.DexCodeItem;
import jeb.api.dex.DexDalvikInstruction;
import jeb.api.ui.JavaView;
import jeb.api.ui.View;

public class WhatsAppPlugin implements IScript {
	
	char[] keys={0x53,0x26,0x55,0x3A,0x8}; // Change keys here

	public String decode_string(String input)
	{
		char [] temp=input.toCharArray();
		for (int i=0;i<input.length();i++)
		{
			temp[i]=(char)(temp[i]^keys[i%5]);
		}
		String output=new String(temp).intern();
		return output;
	}
	
	public void run(JebInstance jeb)
	{
		Dex dex=jeb.getDex();
		JavaView view=(JavaView)jeb.getUI().getView(View.Type.JAVA);
		String methodname=view.getCodePosition().getSignature();
		jeb.print(methodname);
		DexCodeItem code=dex.getMethodData(methodname).getCodeItem();
		if (code==null) return;
		List<DexDalvikInstruction> list=code.getInstructions();
		HashMap<Integer, String> staticstrings = new HashMap<Integer, String>();
		
		for (DexDalvikInstruction insn: list)
		{
			String mnemonic=insn.getMnemonic();
			if (!mnemonic.contains("const-string")) continue;
			long stringindex=insn.getParameters()[1].getValue();
			String cryptstring=dex.getString((int)stringindex);
			if (staticstrings.containsKey(stringindex)==false)
			{
				staticstrings.put((int)stringindex,cryptstring);
			}
		}
		for(int key : staticstrings.keySet())
		{
			String value=decode_string(staticstrings.get(key));
			String str=String.valueOf(key)+": "+value;
			dex.setString(key, value);
			jeb.print(str);
		}
	}
}
